api/internal/target/kusttarget.go

```diff
@@ -260,44 +268,78 @@ func (kt *KustTarget) configureExternalTransformers() ([]resmap.Transformer, err
 // accumulateResources fills the given resourceAccumulator
 // with resources read from the given list of paths.
 func (kt *KustTarget) accumulateResources(
-	ra *accumulator.ResAccumulator, paths []string) error {
+	ra *accumulator.ResAccumulator, paths []string) (*accumulator.ResAccumulator, error) {
 	for _, path := range paths {
 		// try loading resource as file then as base (directory or git repository)
 		if errF := kt.accumulateFile(ra, path); errF != nil {
 			ldr, errL := kt.ldr.New(path)
 			if errL != nil {
-				return fmt.Errorf("accumulateFile %q, loader.New %q", errF, errL)
+				return nil, fmt.Errorf("accumulateFile %q, loader.New %q", errF, errL)
 			}
-			errD := kt.accumulateDirectory(ra, ldr)
+			var errD error
+			ra, errD = kt.accumulateDirectory(ra, ldr, false)
 			if errD != nil {
-				return fmt.Errorf("accumulateFile %q, accumulateDirector: %q", errF, errD)
+				return nil, fmt.Errorf("accumulateFile %q, accumulateDirector: %q", errF, errD)
 			}
 		}
 	}
-	return nil
+	return ra, nil
+}
+
+// accumulateResources fills the given resourceAccumulator
+// with resources read from the given list of paths.
+func (kt *KustTarget) accumulateComponents(
+	ra *accumulator.ResAccumulator, paths []string) (*accumulator.ResAccumulator, error) {
+	for _, path := range paths {
+		// Components always refer to directories
+		ldr, errL := kt.ldr.New(path)
+		if errL != nil {
+			return nil, fmt.Errorf("loader.New %q", errL)
+		}
+		var errD error
+		ra, errD = kt.accumulateDirectory(ra, ldr, true)
+		if errD != nil {
+			return nil, fmt.Errorf("accumulateDirectory: %q", errD)
+		}
+	}
+	return ra, nil
 }
 
 func (kt *KustTarget) accumulateDirectory(
-	ra *accumulator.ResAccumulator, ldr ifc.Loader) error {
+	ra *accumulator.ResAccumulator, ldr ifc.Loader, isComponent bool) (*accumulator.ResAccumulator, error) {
 	defer ldr.Cleanup()
 	subKt := NewKustTarget(
 		ldr, kt.validator, kt.rFactory, kt.tFactory, kt.pLdr)
 	err := subKt.Load()
 	if err != nil {
-		return errors.Wrapf(
+		return nil, errors.Wrapf(
 			err, "couldn't make target for path '%s'", ldr.Root())
 	}
-	subRa, err := subKt.AccumulateTarget()
+	if isComponent && subKt.kustomization.Kind != types.ComponentKind {
+		return nil, fmt.Errorf(
+			"expected kind '%s' for path '%s' but got '%s'", types.ComponentKind, ldr.Root(), subKt.kustomization.Kind)
+	} else if !isComponent && subKt.kustomization.Kind == types.ComponentKind {
+		return nil, fmt.Errorf(
+			"expected kind != '%s' for path '%s'", types.ComponentKind, ldr.Root())
+	}
+
+	var subRa *accumulator.ResAccumulator
+	if subKt.kustomization.Kind == types.ComponentKind {
+		subRa, err = subKt.accumulateTarget(ra)
```

Comment:
Add comments explaining the logic

```go
// components don't create a new accumulator, the kustomization directives are added to the current accumulator
...
// kustomizations create a new accumulator which resolvers the kustomization directives before adding the resulting resources to the current accumulator
```

Reply:
Good idea - I slightly rewrote your suggestion, so does it look ok?